libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc

```diff
@@ -321,32 +321,293 @@ void flystel_prime_field_gadget<ppT, parameters>::generate_r1cs_witness()
+        const std::vector<FieldT> &C,
+        const std::vector<FieldT> &D,
+        const pb_linear_combination_array<FieldT> &X_left,
+        const pb_linear_combination_array<FieldT> &X_right,
+        const pb_variable_array<FieldT> &Y_left,
+        const pb_variable_array<FieldT> &Y_right,
+        const std::string &annotation_prefix)
+    : gadget<libff::Fr<ppT>>(pb, annotation_prefix)
+    , C_const(C)
+    , D_const(D)
+    , X_left_input(X_left)
+    , X_right_input(X_right)
+    , Y_left_output(Y_left)
+    , Y_right_output(Y_right)
+{
+    const libff::Fr<ppT> g = parameters::multiplicative_generator_g;
+    const size_t ncols = NumStateColumns_L;
+
+    // temporary variables (Z_left, Z_right) modified in-place during
+    // the computation from (X_left, X_right) to (Y_left, Y_right)
+    std::vector<linear_combination<libff::Fr<ppT>>> Z_left;
+    std::vector<linear_combination<libff::Fr<ppT>>> Z_right;
+
+    // add constants Z_left[i]+=C[i], Z_right[i]+=D[i]
+    for (size_t i = 0; i < ncols; i++) {
+        Z_left.push_back(X_left[i] + C[i]);
+        Z_right.push_back(X_right[i] + D[i]);
+    }
+
+    if (ncols > 1) {
+        M_matrix = anemoi_permutation_mds<ppT, ncols>(g);
+    } else { // ncols == 1
+        // the MDS matrix for a state with 1 column (L=1) is the same as
+        // for a state with 2 columns (L=2)
+        M_matrix = anemoi_permutation_mds<ppT, 2>(g);
+    }
+
+    // multiply by matrix M
+    if (ncols > 1) {
+        // l > 1:
+        // Z_left  = (zL_0 zL_1 ... zL_{l-1})
+        // Z_right = (zR_0 zR_1 ... zR_{l-1})
+        // Z_left = M Z_left
+        // Z_right = M (Z_right <<< 1)
+        // where (Z_right <<< 1) = (zR_1 ... zR_{l-1} zR_0)
+        Z_left = anemoi_fast_multiply_mds<ppT, NumStateColumns_L>(Z_left, g);
+        std::vector<linear_combination<libff::Fr<ppT>>> Z_right_lrot =
+            anemoi_vector_left_rotate_by_one<ppT>(Z_right);
+        Z_right =
+            anemoi_fast_multiply_mds<ppT, NumStateColumns_L>(Z_right_lrot, g);
+    } else { // ncols == 1
+        // l = 1:
+        // Z_left = zL_0
+        // Z_right = zR_0
+        // Z = Z_left || Z_right
+        // Z = M Z
+        // Z_left  = Z[0]
+        // Z_right = Z[1]
+        assert(Z_left.size() == 1);
+        assert(Z_right.size() == 1);
+        std::vector<linear_combination<libff::Fr<ppT>>> Z;
+        Z.push_back(Z_left[0]);
+        Z.push_back(Z_right[0]);
+        // for L=1 still calling multiply routine for L=2
+        Z = anemoi_fast_multiply_mds<ppT, 2>(Z, g);
+        assert(Z.size() == 2);
+        Z_left.clear();
+        Z_right.clear();
+        Z_left.push_back(Z[0]);
+        Z_right.push_back(Z[1]);
+    }
+
+    // apply layer of L Flystel S-boxes
+    for (size_t i = 0; i < ncols; i++) {
+        flystel_prime_field_gadget<ppT, parameters> H(
+            pb, Z_left[i], Z_right[i], Y_left[i], Y_right[i], "Flystel[i]");
+        Flystel.push_back(H);
```

Comment:
Simpler to use `emplace_back`?

Reply:
Fixed

Reply:
You've just replaced `puh_back` by `emplace_back`, which does not make anything simpler.  The point of `emplace_back` is that you don't have to create an intermediate `flystel_prime_field_gadget` object, you can pass the arguments to the constructor of the element directly into `emplace_back`.

Reply:
Fixed. See #101.